### PR TITLE
router: add support for prefix and suffix header matching

### LINKED ISSUE #3563: Add prefix/suffix match to HeaderMatcher
*Title*: *Support prefix/suffix match in HeaderMatcher*

*Description*:
The current [HeaderMatcher](https://github.com/envoyproxy/envoy/blob/5c7027c34e93fa0cb0574e74757dab637b847255/api/envoy/api/v2/route/route.proto#L910) doesn't support prefix/suffix match. Even though it supports regex match, a dedicated prefix/suffix match still have better performance and much easier to use.

### PR DESCRIPTION
**Pull request by yangminzhu**
Signed-off-by: Yangmin Zhu <ymzhu@google.com>

*Description*:
This adds `prefix` and `suffix` match in HeaderMatcher .

*Risk Level*:
Low

*Testing*:
Added unit tests.

*Docs Changes*:
Inline with proto change.

*Release Notes*:
Noted the new options.

Fixes #3563

### CODE REVIEW

**Comment by yangminzhu** [comment]
@mattklein123 @rodaine PTAL, all tests passed. Thanks!

File: api/envoy/api/v2/route/route.proto [hunk 1]
```diff
@@ -939,6 +939,22 @@ message HeaderMatcher {
     // If specified, header match will be be performed based on whether the header is in the
     // request.
     bool present_match = 7;
+
```
Comment (lizan) [inline api/envoy/api/v2/route/route.proto]: Should we bring `StringMatch` here and potentially deprecate `regex_match` and `exact_match`?

File: api/envoy/api/v2/route/route.proto [hunk 2]
```diff
@@ -939,6 +939,22 @@ message HeaderMatcher {
...
+    // If specified, header match will be performed based on the prefix of the header value.
+    //
+    // Examples:
+    //
+    // * The prefix *abcd* matches the value *abcdxyz*, but not for *abcxyz*.
+    // * The empty prefix is also valid and matches any value.
```
Comment (lizan) [inline api/envoy/api/v2/route/route.proto]: we already have `present_match` above, I don't think empty prefix adds value. (same for suffix_match)

File: api/envoy/api/v2/route/route.proto (hunk 1 above)
Comment (mattklein123) [reply api/envoy/api/v2/route/route.proto]: That would be optimal, though, we have gone through a lot of churn here and TBH changes in this structure are a huge PITA for all users, so I would probably say it's not worth it.

File: api/envoy/api/v2/route/route.proto (hunk 2 above)
Comment (yangminzhu) [reply api/envoy/api/v2/route/route.proto]: Yeah, I just want to make it clear in the comment that an empty prefix/suffix will match to anything which is how people usually would expect for prefix/suffix match.
Do you want to explicitly disallow empty string here? I feel it's no harm to provide more flexibility to user.

Comment (mattklein123) [reply api/envoy/api/v2/route/route.proto]: I would probably disallow it using a constraint. IMO there is no reason to allow multiple ways of doing the same thing within the same config structure.

File: api/envoy/api/v2/route/route.proto (hunk 1 above)
Comment (yangminzhu) [reply api/envoy/api/v2/route/route.proto]: I added a validation to check it's not empty, also updated the comment. Do we also want to check this in code and always return false if the value is empty?

**Pull request merged**